File: apps/android/app/src/main/java/dev/flextranslate/foundation/CloudMediationClient.kt
package dev.flextranslate.foundation

import android.util.Log
import org.json.JSONObject
import java.io.BufferedReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * Запрос, который приложение шлёт НАШЕМУ бэкенду-посреднику для перевода через Gemini Flash.
 * Несёт только *намерение* пользователя (пара языков + текст + id модели), но не credential Gemini.
 * Бэкенд сам подставляет авторизацию, дёргает `generateContent` Gemini на своей стороне и
 * возвращает только текст.
 *
 * Формат провода — см. `docs/design/ws5-gemini-flash.md` §2.3-A.
 */
data class GeminiTranslateRequest(
    val providerId: String,
    val modelId: String,
    val languagePair: String,
    val text: String,
    val thinkingLevel: String,
    val stream: Boolean = false,
)

/**
 * Шов между [GeminiFlashTranslationProvider] и бэкендом, который держит оператор. Реализации —
 * единственное место, которое лезет в сеть ради облачного MT; так провайдер остаётся тестируемым
 * (подсунул фейковый клиент) и инвариант «никаких вшитых ключей» живёт в одном файле.
 *
 * Результат — явный sealed-тип, чтобы провайдер мог честно отразить любой исход в
 * [TranslationResult] (текст успеха, причину отказа бэкенда или сбой транспорта) и НИКОГДА
 * не выдумывал вывод.
 */
interface CloudMediationClient {
    fun translate(request: GeminiTranslateRequest, credential: CloudCredential): Result

    sealed interface Result {
        /** Бэкенд вернул реальный перевод, сделанный Gemini на сервере. */
        data class Ok(val text: String, val modelId: String?) : Result

        /** Бэкенд явно отказал (лимит, политика, safety) с понятным пользователю сообщением. */
        data class Refused(val userReason: String) : Result

        /** Сбой транспорта или парсинга — сеть отвалилась на полпути, кривой ответ, таймаут и т.п. */
        data class Failed(val cause: String) : Result
    }
}

/**
 * Боевой клиент-посредник поверх [HttpURLConnection] (без лишней HTTP-зависимости — приложение и так
 * избегает тяжёлых либ). POST'ит намерение перевода в JSON на настроенный endpoint, кладя в
 * `Authorization` собственный сессионный токен приложения — но **никогда** ключ Gemini, которого у
 * приложения нет.
 *
 * Контракт честности: любой не-2xx, кривое тело или `{ "ok": false }` превращаются в [Refused]/[Failed],
 * а вызывающий показывает честную причину. Выдумать перевод тут нечем.
 */
class HttpCloudMediationClient(
    private val config: GeminiFlashConfig,
) : CloudMediationClient {

    override fun translate(
        request: GeminiTranslateRequest,
        credential: CloudCredential,
    ): CloudMediationClient.Result {
        val endpoint = config.translateEndpoint()
            ?: return CloudMediationClient.Result.Failed("backend endpoint not configured")

        return runCatching { post(endpoint, buildRequestBody(request), credential) }
            .getOrElse { t ->
                Log.w(TAG, "mediated translate failed", t)
                CloudMediationClient.Result.Failed(t.message ?: t.javaClass.simpleName)
            }
    }

    private fun post(
        endpoint: String,
        body: String,
        credential: CloudCredential,
    ): CloudMediationClient.Result {
        val connection = (URL(endpoint).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"
            connectTimeout = config.timeoutMs
            readTimeout = config.timeoutMs
            doOutput = true
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            setRequestProperty("Accept", "application/json")
            // Это сессионная идентичность приложения для бэкенда, а НЕ ключ Gemini API. Ключ Gemini
            // лежит в серверном окружении бэкенда-посредника.
            setRequestProperty("Authorization", "Bearer ${credential.source}")
        }
        return try {
            connection.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            val status = connection.responseCode
            val stream = if (status in 200..299) connection.inputStream else connection.errorStream
            val payload = stream?.bufferedReader()?.use(BufferedReader::readText).orEmpty()
            parseResponse(status, payload)
        } finally {
            connection.disconnect()
        }
    }

    private companion object {
        const val TAG = "CloudMediationClient"
    }
}

/** Собирает JSON-тело запроса перевода. Вынесено отдельно, чтобы форму запроса можно было тестить. */
internal fun buildRequestBody(request: GeminiTranslateRequest): String =
    JSONObject().apply {
        put("providerId", request.providerId)
        put("modelId", request.modelId)
        put("languagePair", request.languagePair)
        put("text", request.text)
        put("thinkingLevel", request.thinkingLevel)
        put("stream", request.stream)
    }.toString()

/**
 * Разбирает ответ бэкенда в честный [CloudMediationClient.Result]. Формы (§2.3-A):
 *  - успех: `{ "ok": true, "text": "...", "modelId": "..." }`
 *  - отказ: `{ "ok": false, "reason": "...", "userMessage": "..." }`
 *  - любой не-2xx или кривое тело → [Failed].
 */
internal fun parseResponse(status: Int, payload: String): CloudMediationClient.Result {
    val json = runCatching { JSONObject(payload) }.getOrNull()
        ?: return CloudMediationClient.Result.Failed("HTTP $status: unparseable response")

    if (status !in 200..299) {
        val message = json.optString("userMessage").ifBlank { "HTTP $status" }
        return CloudMediationClient.Result.Refused(message)
    }
    if (!json.optBoolean("ok", false)) {
        val message = json.optString("userMessage")
            .ifBlank { json.optString("reason") }
            .ifBlank { "Облачный перевод отклонён сервером" }
        return CloudMediationClient.Result.Refused(message)
    }
    val text = json.optString("text")
    if (text.isBlank()) {
        return CloudMediationClient.Result.Failed("backend ok but empty text")
    }
    val modelId = json.optString("modelId").ifBlank { null }
    return CloudMediationClient.Result.Ok(text = text, modelId = modelId)
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/CloudMediationClient.swift
import Foundation

/// Запрос, который приложение шлёт НАШЕМУ backend-посреднику на перевод через Gemini Flash.
/// Несёт только намерение пользователя — ключа Gemini тут нет. Backend сам подставляет
/// авторизацию, дёргает generateContent на своей стороне и возвращает только текст.
///
/// Калька с Android GeminiTranslateRequest.
struct GeminiTranslateRequest: Sendable {
    let providerId: String
    let modelId: String
    let languagePair: String
    let text: String
    let thinkingLevel: String
    let stream: Bool
}

/// Граница между GeminiFlashTranslationProvider и backend, который держит оператор.
/// Результат — явный enum, чтобы провайдер мог честно разложить любой исход на
/// TranslationResult (текст / причина отказа / ошибка транспорта) и НИКОГДА не выдумывал
/// вывод.
///
/// Калька с Android CloudMediationClient.
protocol CloudMediationClient: Sendable {
    func translate(request: GeminiTranslateRequest, credential: CloudCredential) -> CloudMediationResult
}

enum CloudMediationResult: Sendable {
    /// Backend вернул настоящий перевод, сделанный Gemini на серверной стороне.
    case ok(text: String, modelId: String?)
    /// Backend явно отказал с сообщением на языке продукта.
    case refused(String)
    /// Сбой транспорта или парсинга.
    case failed(String)
}

/// Клиент посредника поверх URLSession (без лишних зависимостей).
/// POST-ит намерение-перевод как JSON на настроенный эндпоинт backend, в Authorization
/// кладёт собственный сессионный токен приложения — НИКОГДА не ключ Gemini.
///
/// Калька с Android HttpCloudMediationClient.
final class HttpCloudMediationClient: CloudMediationClient, @unchecked Sendable {

    private let config: GeminiFlashConfig

    init(config: GeminiFlashConfig) {
        self.config = config
    }

    func translate(request: GeminiTranslateRequest, credential: CloudCredential) -> CloudMediationResult {
        guard let endpoint = config.translateEndpoint() else {
            return .failed("backend endpoint not configured")
        }
        guard let url = URL(string: endpoint) else {
            return .failed("invalid backend endpoint URL")
        }
        let body = buildMediatedRequestBody(request: request)
        var urlRequest = URLRequest(url: url, timeoutInterval: config.timeoutSeconds)
        urlRequest.httpMethod = "POST"
        urlRequest.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        urlRequest.setValue("application/json", forHTTPHeaderField: "Accept")
        // Сессионная идентичность приложения для нашего backend — НЕ ключ Gemini.
        urlRequest.setValue("Bearer \(credential.source)", forHTTPHeaderField: "Authorization")
        urlRequest.httpBody = body.data(using: .utf8)

        let semaphore = DispatchSemaphore(value: 0)
        var responseData: Data?
        var httpStatus = 0
        var networkError: String?

        URLSession.shared.dataTask(with: urlRequest) { data, response, error in
            if let error = error {
                networkError = error.localizedDescription
            } else if let http = response as? HTTPURLResponse {
                httpStatus = http.statusCode
                responseData = data
            }
            semaphore.signal()
        }.resume()
        semaphore.wait()

        if let msg = networkError {
            return .failed(msg)
        }
        let payload = responseData.flatMap { String(data: $0, encoding: .utf8) } ?? ""
        return parseMediationResponse(status: httpStatus, payload: payload)
    }
}

/// Собирает JSON-тело запроса к посреднику. Вынесено отдельно, чтобы форму запроса можно было покрыть тестами.
func buildMediatedRequestBody(request: GeminiTranslateRequest) -> String {
    func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "\\", with: "\\\\")
         .replacingOccurrences(of: "\"", with: "\\\"")
         .replacingOccurrences(of: "\n", with: "\\n")
    }
    let pid = escape(request.providerId)
    let mid = escape(request.modelId)
    let lp = escape(request.languagePair)
    let txt = escape(request.text)
    let tl = escape(request.thinkingLevel)
    let st = request.stream ? "true" : "false"
    return "{\"providerId\":\"\(pid)\",\"modelId\":\"\(mid)\",\"languagePair\":\"\(lp)\",\"text\":\"\(txt)\",\"thinkingLevel\":\"\(tl)\",\"stream\":\(st)}"
}

/// Превращает ответ backend в честный CloudMediationResult.
/// Калька с Android parseResponse.
func parseMediationResponse(status: Int, payload: String) -> CloudMediationResult {
    guard let data = payload.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        return .failed("HTTP \(status): unparseable response")
    }

    if !(200...299).contains(status) {
        let message = (json["userMessage"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? "HTTP \(status)"
        return .refused(message)
    }
    let ok = json["ok"] as? Bool ?? false
    if !ok {
        let message = (json["userMessage"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? (json["reason"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            ?? "Cloud translation declined by server"
        return .refused(message)
    }
    guard let text = json["text"] as? String, !text.isEmpty else {
        return .failed("backend ok but empty text")
    }
    let modelId = (json["modelId"] as? String).flatMap { $0.isEmpty ? nil : $0 }
    return .ok(text: text, modelId: modelId)
}
